src/debug_flags.h

```diff
@@ -17,12 +17,20 @@
 
 #include <stdio.h>
 
-#define EXPLAIN(fmt, ...) {                                             \
-  if (g_explaining)                                                     \
-    fprintf(stderr, "ninja explain: " fmt "\n", __VA_ARGS__);           \
+#define EXPLAIN(node, fmt, ...) {                                       \
+  if (g_explaining) {                                                   \
+    char buf[1024];                                                     \
+    snprintf(buf, 1024, fmt, __VA_ARGS__);                              \
+    record_explanation(node, buf);                                      \
```

Comment:
Consider doing the temporary buffer allocation directly in a `record_explanation`, as in:

```
void record_explanation(const Node*, const char* fmt, ...);
```

Reply:
Actually, with a dedicated class and pointed object, you can drop the macro entirely and replace all instances of `EXPLAIN(.....)` with `explanations->Record(node, ....)` which will be clearer.